lib/protoboeuf/codegen.rb

```diff
@@ -280,7 +280,7 @@ def encode_string(field, value_expr, tagged)
           val = #{value_expr}
           if((len = val.bytesize) > 0)
             #{encode_tag_and_length(field, tagged, "len")}
-            buff << val
+            buff << (val.ascii_only? ? val : val.b)
```

Comment:
If the string is pure ASCII we can skip the (1 or 2) allocations, it's debatable whether it's worth it to check though.

Reply:
Probably worth it to check 👍

Reply:
Can you explain why you apppend val as-is if ASCII only, and otherwise append as bytes? Is this to maintain the code range?

Reply:
Could it be faster to concat the string without any checking or calling `.b` and then calling `force_encoding` on buf after?

Reply:
Because of the encoding negociation I explained before.

You normally can't concatenate `UTF-8` into `ASCII-8BIT`, they are different encoding, so concatenating one into the other raise an error.

HOWEVER, they are both marked as "ascii_compatible", which means that if a string with this encoding only contains bytes in the ASCII range (7BIT), then it's allowed to implicitly be considered as ASCII, which allow the concatenation on the first time around.

Here you start with: `"abcdf"(ASCII-8BIT)`, it only contains ASCII so it accepts UTF-8 (invalid but wtv) being appended, and "promote" itself to UTF-8.

On the second call, the buffer is `"abcdf\xFE"(ASCII-8BIT)`, it's no longer ASCII only, so the encoding negociation refuse to consider it as just ASCII and to promote it.

Reply:
Yes, that's the big advantage of `append_as_bytes`, and the whole reason behind it.